### PR TITLE
Fix database count, option to disable counting

### PR DESCRIPTION
**Pull request by pxpm**
Mainly this fixes:

1- the total count was always done from model query (Select *), so even if ou apply a scope in your query the total count would also display the ones that the scope discards.
**Example:** 100 users in db, 20 are active. If you apply to the panel: `$this->crud->addClause('active')`, it would still show the total 100 instead of 20 (notice it's a panel query setting like model scopes not filters). 

2 - couldn't handle much data without crashing :-|
At the moment I am working with 500k records with load times under a second depending on db config and the amount of columns needed in query.
-Best times achieved with: `primary` and `unique` keys set on a `bigInteger` as the model key. 

3 - possiblity to don't  use the count functionality at all. 
Even if super optimized, counting 2bill rows is still **counting 2bill rows**, some time will be spent on this operation. If this total/filtered information is not relevant to you, it's better to disable it completely. 

I am pretty sure this needs some docs and probably some refactoring, but I am super curious about the times you will get on your test bench scenario @tabacitu so I put it together to work I think in 90%+ of scenarions (haven't found one failing yet .. but) 
Please be advised about what I told about setting the correct indexes on the columns that are searchable/filterable and having `unique` on the primary key really helps!

### CODE REVIEW

**Comment by tabacitu** [comment]
oh-my-god-reaction-gif-by-friends

Pedro. PEDROOO! THIS FREAKING WORKS! IT WORKS!!!

I have two tables who haven't worked for YEARS. One with 1.2M entries, one with 1.7M. I just switched to this branch and... they work. They work again 🎉 This is freaking awesome, Pedro! 

Let's clean it up, please.

File: src/resources/views/crud/inc/datatables_logic.blade.php [hunk 1]
```diff
@@ -246,7 +248,10 @@ functionsToRunOnDataTablesDrawEvent: [],
           searching: @json($crud->getOperationSetting('searchableTable') ?? true),
           ajax: {
               "url": "{!! url($crud->route.'/search').'?'.Request::getQueryString() !!}",
-              "type": "POST"
+              "type": "POST",
+              "data": {
+                "unfilteredQueryCount": "{{$crud->getOperationSetting('unfilteredQueryCount') ?? false}}"
+            },
```
Comment (tabacitu) [inline src/resources/views/crud/inc/datatables_logic.blade.php:254]: Do we actually need to include `unfilteredQueryCount` as a parameter upon each search request? I don't understand why...  I think this is only needed if we let the user change the count on the front-end... which... we don't... right?

Otherwise, `unfilteredQueryCount` is in our control in the endpoint too... so why not define true/false only there?

File: src/app/Http/Controllers/Operations/ListOperation.php [hunk 2]
```diff
@@ -71,17 +71,19 @@ public function search()
     {
         $this->crud->hasAccessOrFail('list');
 
+        $showEntryCount = $this->crud->getOperationSetting('showEntryCount');
+
+        $this->crud->setOperationSetting('unfilteredQueryCount', request('unfilteredQueryCount') ?? $this->crud->getOperationSetting('unfilteredQueryCount'));
+
+        $totalRows = ! $showEntryCount ? 0 : $this->crud->getOperationSetting('unfilteredQueryCount') ?? $this->crud->getQueryCount();
```
Comment (tabacitu) [inline src/app/Http/Controllers/Operations/ListOperation.php:78]: Messy. Let's clean it up.

**Review by tabacitu — COMMENTED** [review]
@pxpm could you please give this a polish? Make it pretty and easy to understand please 🙏 It's worth it, we're gonna merge this as soon as it's ready. 

fox-gif-by-the-masked-singer

File: src/resources/views/crud/inc/datatables_logic.blade.php (hunk 1 above)
Comment (pxpm) [reply src/resources/views/crud/inc/datatables_logic.blade.php:254]: When the search ajax query is performed we already have the page loaded so the query to get the `unfilteredQueryCount` already had run.
If we send it  for the ajax endpoint we don't need to "re-do" this `unfilteredQueryCount` and only do the `filteredQueryCount` after applying what was sent to the `search()` endpoint, otherwise we would calculate the `filteredQueryCount` when defining the panel filters

Note that: `adding filters to the crud applies the query constrain` so we need to get the `unfilteredQuery` before applying any filter or any seach, only account for the base $crud->query after the `setups` definitions in the CrudController.

File: src/app/Library/CrudPanel/CrudFilter.php [hunk 3]
```diff
@@ -101,6 +101,12 @@ public function wasNotApplied()
      */
     public function apply($input = null)
     {
+        // before applying filters, store the base query count (query before any filters are applied)
+        $showEntryCount = $this->crud()->getOperationSetting('showEntryCount');
+        $unfilteredQueryCount = request('unfilteredQueryCount') ?? $this->crud()->getOperationSetting('unfilteredQueryCount');
+        $unfilteredQueryCount = ! $showEntryCount ? 0 : ($unfilteredQueryCount ?? $this->crud()->getQueryCount());
+        $this->crud()->setOperationSetting('unfilteredQueryCount', $unfilteredQueryCount);
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/CrudFilter.php:108]: This is still a bit messy:

(1) Why have the `$showEntryCount` variable in the first place?
- first, we calculate `$showEntryCount`
- then in the next line, we don't use it
- then in the third line, we do use it, but only there
- => we could have just used `$this->crud()->getOperationSetting('showEntryCount')` inline, right? That would have been cleaner.

(2) Even after the above, seems like we're doing a lot of calculation, all on top of `$unfilteredQueryCount`. So then why not simplify all this bit of logic into a function? Even if we did it in a closure, I think it would read easier:

```suggestion
        // store the base query count, before applying the filters
        $this->crud()->setOperationSetting('unfilteredQueryCount', (function () {
            if (!$this->crud()->getOperationSetting('showEntryCount')) {
                return 0;
            }

            return $this->crud()->getRequest()->input('unfilteredQueryCount') ??
                   $this->crud()->getOperationSetting('unfilteredQueryCount') ??
                   $this->crud()->getQueryCount();
        })());
```

Wouldn't you agree?

Comment (tabacitu) [reply src/app/Library/CrudPanel/CrudFilter.php:108]: (3) Please notice that above I've used `$this->crud()->getRequest()->input('unfilteredQueryCount')` instead of `request('unfilteredQueryCount')`. Let's make a note that we should always use the CRUD request, not the helper. That's what we do below. Just in case it's been overriden, we should use the same thing across the file.

(4) Now that we have this bit of logic separated into a closure, we can EASILY re-use it if we ever need it again, by just moving it to a method on the `CrudPanel` object.

File: src/app/Http/Controllers/Operations/ListOperation.php [hunk 4]
```diff
@@ -71,17 +71,23 @@ public function search()
...
 
+        // should the `Showing 1 of 10 ... ` be displayed ? When it's not displayed we wouldn't do the counting calculations.
+        $showEntryCount = $this->crud->getOperationSetting('showEntryCount');
+
+        // how many total entries there is applying the controller/model constrains, before any filtering or search
+        $unfilteredQueryCount = request('unfilteredQueryCount') ?? $this->crud->getOperationSetting('unfilteredQueryCount');
+
+        $totalRows = ! $showEntryCount ? 0 : $unfilteredQueryCount ?? $this->crud->getQueryCount();
+
+        $this->crud->setOperationSetting('unfilteredQueryCount', $totalRows);
```
Comment (tabacitu) [inline src/app/Http/Controllers/Operations/ListOperation.php:82]: Same feedback as in the `CrudFilter` file. Aaaand... it looks like it's the exact same calculation, isn't it? In that case, it probably makes sense to move that logic to its own function, something like `$this->crud->setUnfilteredQueryCount()` or `$this->crud->setUnfilteredQueryCount()`. Right?

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 5]
```diff
@@ -148,4 +150,105 @@ public function orderByWithPrefix($column_name, $column_direction = 'ASC')
 
         return $this->query->orderBy($column_name, $column_direction);
     }
+
+    /**
+     * return the nested query columns.
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Query.php:155]:
```suggestion
     * Return the nested query columns.
```

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 6]
```diff
@@ -148,4 +150,105 @@ public function orderByWithPrefix($column_name, $column_direction = 'ASC')
...
+    }
+
+    /**
+     * we want to select the minimum possible columns respecting the clauses in the query, so that the count is accurate.
+     * for that to happen we will traverse the query `wheres` (Basic, Exists, Nested or Column) to get the correct
+     * column that we need to select in the main model for that "sub query" to work.
+     *
+     * For example a base query of: `SELECT * FROM table WHERE (someColumn, smtValue)`, we would return only the `someColumn`
+     * with the objective of replacing the `*` for the specific columns needed, avoiding the selection of
+     * columns that would not have impact in the counting process.
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Query.php:172]: Freaking EXCELLENT explanation here Pedro. 👏👏👏 Just one nitpick:

```suggestion
     * We want to select the minimum possible columns respecting the clauses in the query, so that the count is accurate.
     * for that to happen we will traverse the query `wheres` (Basic, Exists, Nested or Column) to get the correct
     * column that we need to select in the main model for that "sub query" to work.
     *
     * For example a base query of: `SELECT * FROM table WHERE (someColumn, smtValue)`, we would return only the `someColumn`
     * with the objective of replacing the `*` for the specific columns needed, avoiding the selection of
     * columns that would not have impact in the counting process.
```

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 7]
```diff
@@ -148,4 +150,105 @@ public function orderByWithPrefix($column_name, $column_direction = 'ASC')
...
+
+                break;
+            }
+        }
+
+        return $wheresColumns;
+    }
+
+    /**
+     * count of available entries from the current crud query.
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Query.php:215]:
```suggestion
     * Do a separate query to get the total number of entries, in an optimized way.
```

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 8]
```diff
@@ -148,4 +150,105 @@ public function orderByWithPrefix($column_name, $column_direction = 'ASC')
...
         return $this->query->orderBy($column_name, $column_direction);
     }
+
+    /**
+     * return the nested query columns.
+     *
+     * @param  \Illuminate\Database\Eloquent\Builder  $query
+     * @return array
+     */
+    public function getNestedQueryColumns($query)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Query.php:160]: Does this really need to be `public`?

```suggestion
    private function getNestedQueryColumns($query)
```

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 9]
```diff
@@ -148,4 +150,105 @@ public function orderByWithPrefix($column_name, $column_direction = 'ASC')
...
+     *
+     * For example a base query of: `SELECT * FROM table WHERE (someColumn, smtValue)`, we would return only the `someColumn`
+     * with the objective of replacing the `*` for the specific columns needed, avoiding the selection of
+     * columns that would not have impact in the counting process.
+     *
+     * @param  \Illuminate\Database\Eloquent\Builder  $query
+     * @param  bool  $nested  used to prevent multiple level nesting as we only need the first level columns
+     * @return array
+     */
+    public function getQueryColumnsFromWheres($query, $nested = false)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Query.php:178]: Does the dev ever need to call this manually?

```suggestion
    private function getQueryColumnsFromWheres($query, $nested = false)
```

**Review by tabacitu — CHANGES_REQUESTED** [review]
Pedro, I couldn't focus on the business logic, there are still some hygiene improvements to do here, that kept distracting me. Let's finish these first, then I can take a look at the logic too 🙏

**Comment by pxpm** [comment]
@tabacitu if I were to rate your feedback I would say 10/10! GJ. 👍 

I've refactored according to it, and I am happy how it looks right now, do you have any other suggestion ? 

Cheers

**Comment by pxpm** [comment]
heads up, updated this PR with a new rule:
- Don't optimize the column selection from query when developer uses raw sql. Otherwise we would need to parse that sql ... and we know that hole is deeper than mariana trench!

**Comment by jorgetwgroup** [comment]
Hi there @pxpm i hope you are well.

First, im very sorry because im still learning about deep side of Backpack, then i dont handle every single detail, but i try my best to test PR's

I say this because i try to disable locally counting with this:

setup()

$this->crud->showEntryCount(false);
$this->crud->disableShowEntryCount();
$this->crud->setOperationSetting('showEntryCount', false);
$this->crud->set('showEntryCount', false);
$this->crud->setShowEntryCount(false);

setupListOperation()

$this->crud->disableShowEntryCount();
$this->crud->setShowEntryCount(false);

No one work, i just need to go to vendor/backpack/crud/src/config/backpack/operations/list.php and disable globally.

First i make a try with main branch, i add 2,293,760 entries to monster table (because have many columns).

As you see, works, but took like +50 sec to handle the count and list.

Screen Shot 2022-08-05 at 12 40 54

Now i go to optionally-show-datatables-info branch, took like 13 sec

Screen Shot 2022-08-05 at 12 42 26

I disable 'showEntryCount' => false, took like 4 sec, can i say how amazing work you do here...im impressed

Screen Shot 2022-08-05 at 12 43 28

I think here my review end, but...

Here is my "grain of sand"...i have a lot of experience with SQL, to be honest i feel more comfortable with Raw than ORM 🙈 i try few scenarios with monsters table and i think i can give to you my idea for this PR to get time similar to 4 sec (best timing) and dont lose count rows...here we go

select count(1) from monsters -> 21.360s
select count(id) from monsters -> 20.700s
select count(*) from monsters -> 20.530s
select count(checkbox) from monsters -> 25.865s

Now let do the magic, i look into table structure and find the smaller column "checkbox (tinyint, 1)"

ALTER TABLE monsters ADD INDEX (checkbox);

select count(checkbox) from monsters -> 0.554s
select count(checkbox) from monsters where checkbox = 1 -> 0.197s

My proposal is add a function who receive a parameter, this parameter will be the optimal column to ask for count, of course this column need to be small and index, and you will get 4 sec. list + count

Let me know if you need any other detail.

Cheers.

**Comment by jorgetwgroup** [comment]
Hi @pxpm

I tried as you say, it is work, but not as expected.

Normal, showing count:
screenshot-backpack test-2022 09 01-12_03_34

Disable count:
screenshot-backpack test-2022 09 01-12_04_06

As you see, this disable count, reduce time, but dont show pagination, then i can not move to other pages.

I tried with all main branches, just use "optionally-show-datatables-info" in CRUD.

Let me know if you need something else.

Cheers.

**Comment by pxpm** [comment]
Hey @jorgetwgroup you are rigth.
I totally forgot about the pagination when disabling the entries count. 

I've fixed that.

Let me know if you spot something else that I've missed. 

Cheers

**Comment by jorgetwgroup** [comment]
Hi @pxpm is working normally now:

https://recordit.co/d1coBYtjJt

Ready to merge.

Cheers.

**Review by tabacitu — CHANGES_REQUESTED** [review]

File: src/app/Http/Controllers/Operations/ListOperation.php [hunk 10]
```diff
@@ -71,32 +71,41 @@ public function search()
     {
         $this->crud->hasAccessOrFail('list');
 
+        // how many total entries there is applying the controller/model constrains, before any filtering or search
+        $this->crud->setOperationSetting('unfilteredQueryCount', $this->crud->getUnfilteredQueryCount());
+
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/app/Http/Controllers/Operations/ListOperation.php:76]: This needs to be placed before filters/search are applied. :|

File: src/app/Http/Controllers/Operations/ListOperation.php [hunk 11]
```diff
@@ -71,32 +71,44 @@ public function search()
     {
         $this->crud->hasAccessOrFail('list');
 
+        // how many total entries there is applying the controller/model constrains, before any filtering or search
+        $this->crud->setOperationSetting('totalEntryCount', $this->crud->getTotalEntryCount());
+
```
Comment (tabacitu) [inline src/app/Http/Controllers/Operations/ListOperation.php:76]: Any reason to do this HERE, at the top of the function @pxpm and not lower, where we calculate the totals? I don't see it used until then anyway, and it's odd to read something like:
- calculate total entry count
- do something else
- do another thing
- yet another thing
- now let's continue calculating the total entry count

Comment (tabacitu) [reply src/app/Http/Controllers/Operations/ListOperation.php:76]: Fixed in https://github.com/Laravel-Backpack/CRUD/pull/4695 - please review & merge if you think it's ok @pxpm

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 12]
```diff
@@ -151,4 +152,150 @@ public function orderByWithPrefix($column_name, $column_direction = 'ASC')
...
+     */
+    public function getTotalEntryCount()
+    {
+        if (! $this->getOperationSetting('showEntryCount')) {
+            return 0;
+        }
+
+        return  $this->getRequest()->input('totalEntryCount') ??
+                $this->getOperationSetting('totalEntryCount') ??
+                $this->getCurrentEntryCount();
```
[the remark this answers is not in the document]
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Query.php:232]: Treated in https://github.com/Laravel-Backpack/CRUD/pull/4695 - please review and merge if you think it's ok @pxpm

File: src/app/Library/CrudPanel/CrudFilter.php [hunk 13]
```diff
@@ -101,6 +101,9 @@ public function wasNotApplied()
      */
     public function apply($input = null)
     {
+        // before applying filters, store the base query count (query before any filters are applied)
+        $this->crud()->setOperationSetting('totalEntryCount', $this->crud()->getTotalEntryCount());
+
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/CrudFilter.php:106]: Hmmm... why is this needed here @pxpm ? Seems out of place.

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 14]
```diff
@@ -151,4 +152,150 @@ public function orderByWithPrefix($column_name, $column_direction = 'ASC')
...
+
+    /**
+     * Return the current query count.
+     *
+     * @return int
+     */
+    public function getCurrentEntryCount()
+    {
+        return $this->getEntryCount();
+    }
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Query.php:243]: I see this method doesn't really do anything other than call a private method. Soo... do we really need it?

**Review by tabacitu — CHANGES_REQUESTED** [review]
Hey @pxpm , I just finished a proper test and review of this, and I'm super-excited to merge it. I've pushed a few small changes already, the ones I knew you'd be ok with (name changes, doc block changes, things like that).

I do have a few more questions and suggestions, please help me clear them up.

Excellent work on this one, can't wait to merge.
Cheers!

File: src/app/Library/CrudPanel/CrudFilter.php (hunk 13 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/CrudFilter.php:106]: (I think you've already explained it to me before 😅, but I forgot 🤦‍♂️)

File: src/resources/views/crud/inc/datatables_logic.blade.php [hunk 15]
```diff
@@ -139,7 +139,7 @@ functionsToRunOnDataTablesDrawEvent: [],
         localStorage.setItem('{{ Str::slug($crud->getRoute()) }}_list_url', newUrl);
       },
       dataTableConfiguration: {
```
Comment (tabacitu) [inline src/resources/views/crud/inc/datatables_logic.blade.php:141]: Question: what happens for someone who's published their `datatables_logic.blade.php` file?
- they do `composer update`
- they get the new PHP files
- they do NOT get this new blade file

In this case... does the list operation break for them?

File: src/app/Http/Controllers/Operations/ListOperation.php (hunk 11 above)
Comment (pxpm) [reply src/app/Http/Controllers/Operations/ListOperation.php:76]: I replied to this question in https://github.com/Laravel-Backpack/CRUD/pull/4695#discussion_r986689198

File: src/app/Library/CrudPanel/CrudFilter.php (hunk 13 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/CrudFilter.php:106]: I replied to this question here https://github.com/Laravel-Backpack/CRUD/pull/4695#discussion_r986689198

File: src/app/Library/CrudPanel/Traits/Query.php (hunk 14 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Query.php:243]: The reason for this acessor method is for keeping `getEntryCount` private, and allow `ListOperation` to call it.

File: src/resources/views/crud/inc/datatables_logic.blade.php [hunk 16]
```diff
@@ -257,10 +257,16 @@ functionsToRunOnDataTablesDrawEvent: [],
...
+          @if($crud->getOperationSetting('showEntryCount') === false)
+            pagingType: "simple",
+          @endif
           searching: @json($crud->getOperationSetting('searchableTable') ?? true),
           ajax: {
               "url": "{!! url($crud->route.'/search').'?'.Request::getQueryString() !!}",
-              "type": "POST"
+              "type": "POST",
+              "data": {
+                "unfilteredQueryCount": "{{$crud->getOperationSetting('unfilteredQueryCount') ?? false}}"
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/resources/views/crud/inc/datatables_logic.blade.php]: Totally agree with the naming! :+1:

File: src/resources/views/crud/inc/datatables_logic.blade.php (hunk 15 above)
Comment (pxpm) [reply src/resources/views/crud/inc/datatables_logic.blade.php:141]: Very interesting question, I haven't thought about it. 

I've just tested it, pulled this branch, and replaced this branch `datatables_logic` blade file with the current version in crud.

Everything works normally unless developer tries to set `showEntryCount => false`. In that case the pagination and the entry count get's messy. 

This branch in it's current state: 
image

This branch with datatables logic replaced:
image

To sum up, any project will work as before ... but if they try to set `showEntryCount` without updating `datatables_logic` it will fail. 
Could this be solved by documentation ? I don't think we can consider this a BC if it does not break the "default" behaviour, and we are not forcing it, developer need to opt-in for `showEntryCount => false`.

Let me know what you think.

Cheers

File: src/app/Library/CrudPanel/CrudFilter.php (hunk 13 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/CrudFilter.php:106]: Hmm... ok... I understand why it's in ListOperation before we call `$this->crud->applyUnappliedFilters();` then... but... why is it here too? 

If we call it there... hasn't this already been done, why call it here too?

File: src/app/Library/CrudPanel/Traits/Query.php (hunk 14 above)
Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Query.php:243]: I understand that. And I would have liked for `getEntryCount()` to be `private` - would have allowed us to change it later. 

But if the only thing a `public` method does is call a `private` method... then that `private` method isn't actually safe at all... it's indirectly callable. Making it `private` and creating an alias for it... doesn't help at all. At least I don't see how.

Comment (tabacitu) [reply src/app/Library/CrudPanel/Traits/Query.php:243]: So my question then becomes... why does `ListOperation` need to call this method directly? `private`/`public`, doesn't matter. Why doesn't `ListOperation` call `getTotalEntryCount()` which
- (1) is already `public`
- (2) loads the count from cache if it exists

?

File: src/resources/views/crud/inc/datatables_logic.blade.php (hunk 15 above)
Comment (tabacitu) [reply src/resources/views/crud/inc/datatables_logic.blade.php:141]: EXCELLEEEEENT! We can consider it a non-breaking change then. Yeah, we can easily solve this with documentation - where we document `showEntryCount` we just tell people to double-check the blade view isn't published 🎉

File: src/app/Library/CrudPanel/Traits/Query.php (hunk 14 above)
Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Query.php:243]: I kind of agree, maybe `protected` would be the best fit here then ? So it could be called by descendants and parents of the class, but not from "outsiders" of the class ? 

```php
class MyPanel extends CrudPanel {
// private getEntryCount can NOT be called
// protected getEntryCount can be called
}

//outsiders
$x = new CrudPanel();
$x->getEntryCount() // will fail in protected and private
```

I think in the end this is a public method as it is a method that is part of the public api (called by list operation, but can be called in other places), I just wanted to put it behind some sketchy visibility setting so our lifes get's easier if we ever neeed to change the method.

Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Query.php:243]: > So my question then becomes... why does `ListOperation` need to call this method directly? `private`/`public`, doesn't matter. Why doesn't `ListOperation` call `getTotalEntryCount()` which
> 
> * (1) is already `public`
> * (2) loads the count from cache if it exists
> 
> ?

`getCurrentEntryCount` it's not the same as `getTotalEntryCount`. I am getting confused by the naming, let me settle the ideas and I will check back here later.

Comment (pxpm) [reply src/app/Library/CrudPanel/Traits/Query.php]: @tabacitu heads up, I did some changes.
- Removed request from the CrudPanel and pass it as a parameter to the function.
- renamed `getCurrentEntryCount` to `performQueryEntryCount`. 

I think this renaming, even if it's not the best name yet, may better describe what is happening in the `search()` function lifecycle. 

I think the name was misleading on what was happening, it made sense only in my head I guess. 

Please review @tabacitu  and let me know if it's now more clear of what's happening and why we don't call `getTotalEntryCount`.

Cheers

File: src/app/Http/Controllers/Operations/ListOperation.php [hunk 17]
```diff
@@ -72,7 +72,7 @@ public function search()
         $this->crud->hasAccessOrFail('list');
 
         // how many total entries there is applying the controller/model constrains, before any filtering or search
-        $this->crud->setOperationSetting('totalEntryCount', $this->crud->getTotalEntryCount());
```
Comment (tabacitu) [inline src/app/Http/Controllers/Operations/ListOperation.php]: Hmmm I still don't think we should pass the request as a parameter here... We can do:

```php
$totalEntryCount = request()->has('totalEntryCount') ? request('totalEntryCount') : $this->crud->getTotalEntryCount();
$this->crud->setOperationSetting('totalEntryCount', $totalEntryCount);
```

Why bloat up `getTotalEntryCount()` with working with the request? I see that method as a CrudPanel method, so it's got nothing to do with the request, it should just answers what's inside CrudPanel, as a setting... otherwise calculate it, store it as the setting, then return it... right?

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 18]
```diff
@@ -219,25 +220,26 @@ private function getQueryColumnsFromWheres($query, $nested = false)
...
     }
 
     /**
-     * Return the current query count.
+     * Runs the query count against the current crud panel query
      *
      * @return int
      */
-    public function getCurrentEntryCount()
+    public function performQueryEntryCount()
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Query.php:242]: I think this name is deceiving... 
If it returns something, it's not a `perform`, it's a `get`... It's a `function`, not a `method`.

I still think this function doesn't make sense to exist... let's just use `getEntryCount()` and that's it (`protected` as you made it here). Can't we do that?

**Review by tabacitu — CHANGES_REQUESTED** [review]

File: src/app/Library/CrudPanel/Traits/Query.php [hunk 19]
```diff
@@ -151,4 +182,150 @@ public function orderByWithPrefix($column_name, $column_direction = 'ASC')
...
+        return $this->getCountFromQuery($this->query);
+    }
+
+    /**
+     * Do a separate query to get the total number of entries, in an optimized way.
+     *
+     * @param  Builder  $query
+     * @return int
+     */
+    protected function getCountFromQuery(Builder $query)
```
Comment (tabacitu) [inline src/app/Library/CrudPanel/Traits/Query.php:280]:
```suggestion
    private function getCountFromQuery(Builder $query)
```

**Comment by tabacitu** [comment]
OMG OMG OMG IT'S MERGED!!! 🤯

**Comment by tabacitu** [comment]
Pedro, correct me if I'm wrong... but we forgot to add documentation for this... I've just added it here - https://backpackforlaravel.com/docs/5.x/crud-operation-list-entries#large-tables-millions-of-entries

Bad maintainers, bad! 😅

Please submit a PR/smth if you feel we should rephrase, or mention it somewhere else.

File: src/app/Http/Controllers/Operations/ListOperation.php [hunk 20]
```diff
@@ -73,30 +73,41 @@ public function search()
...
         }
         // overwrite any order set in the setup() method with the datatables order
         $this->crud->applyDatatableOrder();
 
         $entries = $this->crud->getEntries();
 
-        return $this->crud->getEntriesAsJsonForDatatables($entries, $totalRows, $filteredRows, $startIndex);
+        // if show entry count is disabled we use the "simplePagination" technique to move between pages.
+        if ($this->crud->getOperationSetting('showEntryCount')) {
+            $totalEntryCount = (int) (request()->get('totalEntryCount') ?? $this->crud->getTotalQueryCount());
```
Comment (basselhossam) [inline src/app/Http/Controllers/Operations/ListOperation.php:100]: @pxpm  we need to check if the value is empty instead of null only in case of disabling ConvertEmptyStringsToNull
